source/RenderBuffer.cpp

```diff
@@ -0,0 +1,299 @@
+
+	glEnableVertexAttribArray(shader.Attrib("vert"));
+	glVertexAttribPointer(shader.Attrib("vert"), 2, GL_FLOAT, GL_FALSE, 2 * sizeof(GLfloat), nullptr);
+
+	// unbind the VBO and VAO
+	glBindBuffer(GL_ARRAY_BUFFER, 0);
+	glBindVertexArray(0);
+}
+
+
+
+// Create a texture of the given size that can be used as a render target
+RenderBuffer::RenderBuffer(const Point& dimensions):
+	size(dimensions)
+{
+	// Generate a framebuffer, and bind it.
+	glGenFramebuffers(1, &framebuffer);
+	glBindFramebuffer(GL_FRAMEBUFFER, framebuffer);
+
+	// Generate the texture.
+	glGenTextures(1, &texid);
+	glBindTexture(GL_TEXTURE_2D, texid);
+
+	// Use linear interpolation and no wrapping.
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+
+	// Attach a blank image to the texture.
+	const Point scaledSize = size * Screen::Zoom() / 100;
+	glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, scaledSize.X(), scaledSize.Y(), 0, GL_RGBA, GL_UNSIGNED_BYTE, nullptr);
+
+	// Attach the texture to the frame buffer.
+	glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, texid, 0);
+	GLenum draw_buffers[] = {GL_COLOR_ATTACHMENT0};
+	glDrawBuffers(1, draw_buffers);
+
+	if(glCheckFramebufferStatus(GL_FRAMEBUFFER) != GL_FRAMEBUFFER_COMPLETE)
+	{
+		Logger::LogError("Failed to initialize framebuffer for RenderBuffer");
+	}
+
+
+	glBindTexture(GL_TEXTURE_2D, 0);
+	glBindFramebuffer(GL_FRAMEBUFFER, 0);
+
+	// default to the current viewport size at the time of construction
+	glGetIntegerv(GL_VIEWPORT, last_viewport);
+}
+
+
+
+// Destructor. Frees the texture and renderbuffers.
+RenderBuffer::~RenderBuffer()
+{
+	glDeleteTextures(1, &texid);
+	glDeleteFramebuffers(1, &framebuffer);
+}
+
+
+
+// Turn this buffer on as a render target. The render target is restored if
+// the Activation object goes out of scope.
+RenderBuffer::RenderTargetGuard RenderBuffer::SetTarget()
+{
+	// NOTE: These glGets can cause an unwanted state synchronization that can
+	//       cause performance problems. The only real reason we might want this
+	//       is if we are nesting render buffers. If only one framebuffer is
+	//       enabled at a time, then we can just reset the buffer to 0 when we
+	//       are done.
+	glGetIntegerv(GL_FRAMEBUFFER_BINDING, reinterpret_cast<int*>(&last_framebuffer));
+	glGetIntegerv(GL_VIEWPORT, last_viewport);
+	glBindFramebuffer(GL_FRAMEBUFFER, framebuffer);
+
+	const Point scaledSize = size * Screen::Zoom() / 100;
+	glViewport(0, 0, scaledSize.X(), scaledSize.Y());
+
+	static const float CLEAR[] = {0, 0, 0, 0};
+	glClearBufferfv(GL_COLOR, 0, CLEAR);
+
+	return RenderTargetGuard(*this, size.X(), size.Y());
+}
+
+
+
+// Reset the render target and viewport to the original settings.
+void RenderBuffer::Deactivate()
```

Comment:
this should be a static function in my opinion

Reply:
Right now the design allows for nested buffers, which means the viewport and the framebuffer need to be local variables. We can get rid of that design if you want, but then I need to rewrite portions of the code to use static variables.

Reply:
Hmmm in that case it would be nice to have a static function that just takes you back to the main framebuffer

Reply:
How would a static invocation interact with the non-static version? 

What use case are you trying to support? 

My initial thought is that the static invocation would need to unwind the render buffer stack, which would mean that the stack would have to be available from a static context. This would mean an explicit static vector used as a stack as opposed to the implicit stack of nested RAII objects maintained by the compiler.